File: extra/common/recv_file.fs
\ Copyright (c) 2024 Travis Bemann
\
\ Permission is hereby granted, free of charge, to any person obtaining a copy
\ of this software and associated documentation files (the "Software"), to deal
\ in the Software without restriction, including without limitation the rights
\ to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
\ copies of the Software, and to permit persons to whom the Software is
\ furnished to do so, subject to the following conditions:
\ 
\ The above copyright notice and this permission notice shall be included in
\ all copies or substantial portions of the Software.
\ 
\ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
\ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
\ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
\ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
\ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
\ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
\ SOFTWARE.

begin-module recv-file

  oo import
  systick import
  fat32 import
  crc32 import
  base64 import
  
  begin-module recv-file-internal

    \ The size of a received packet body
    512 constant packet-body-size

    \ Send packets done
    0 constant send-packet-done

    \ Sent the next packet
    1 constant send-packet-next

    \ Resent the current packet
    2 constant send-packet-resend
    
    \ Successfully received packet
    3 constant recv-packet-ack

    \ Unsuccessfully received packet
    4 constant recv-packet-nak

    \ Receive packets failed
    5 constant recv-packet-fail

    \ Send/receive Base64-encoded packet buffer size - do not change this
    700 constant base64-packet-buffer-size

    \ Send/receive Base64-encoded packet buffer
    base64-packet-buffer-size buffer: base64-packet-buffer

    \ Send/receive non-Base64-encoded packet buffer size - do not change this
    packet-body-size 12 + constant packet-buffer-size
    
    \ Send/receive non-Base64-encoded packet buffer
    packet-buffer-size aligned-buffer: packet-buffer

    \ Packet body received length
    0 value packet-body-len
    
    \ Packet body buffer
    packet-body-size buffer: packet-body

    \ The file being written to
    <fat32-file> class-size aligned-buffer: my-file

    \ Receive data into a buffer
    : recv-data ( -- )
      base64-packet-buffer-size 0 do
        key base64-packet-buffer i + c!
      loop
    ;
    
    \ Receive a packet
    : recv-packet ( -- valid? )
      recv-data
      base64-packet-buffer base64-packet-buffer-size
      packet-buffer packet-buffer-size
      decode-base64 nip
    ;

    \ Send a packet
    : send-packet ( message -- )
      packet-buffer !
      0 packet-buffer cell+ !
      packet-buffer 2 cells generate-crc32 packet-buffer 2 cells + !
      packet-buffer 3 cells
      base64-packet-buffer base64-packet-buffer-size
      encode-base64
      base64-packet-buffer swap type
    ;

    \ Handle a packet
    : handle-packet ( -- done? valid? )
      false true { done? valid? }
      recv-packet
      packet-buffer packet-buffer-size cell - generate-crc32
      packet-buffer packet-buffer-size cell - + @ = and if
        packet-buffer cell+ @ { new-packet-body-len }
        new-packet-body-len packet-body-size u> if
          recv-packet-fail send-packet true false exit
          true to done?
          false to valid?
        else
          packet-buffer @ case
            send-packet-done of
              packet-body packet-body-len my-file write-file drop
              true to done?
            endof
            send-packet-next of
              packet-body packet-body-len my-file write-file drop
              new-packet-body-len to packet-body-len
              packet-buffer 2 cells + packet-body packet-body-len move
            endof
            send-packet-resend of
              new-packet-body-len to packet-body-len
              packet-buffer 2 cells + packet-body packet-body-len move
            endof
            false to valid?
          endcase
        then
      else
        false to valid?
      then
      done? valid?
    ;

    \ Transfer data
    : transfer-data ( -- valid? )
      0 to packet-body-len
      false { done? }
      begin done? not while
        ['] handle-packet try { exception }
        exception 0= if
          swap to done?
          if
            recv-packet-ack send-packet
            done? if my-file file-fs@ flush then
          else
            recv-packet-nak send-packet
          then
        else
          recv-packet-fail send-packet
          exception ?raise
        then
      repeat
    ;

    \ Open or create a file
    : open-or-create-file ( path-addr path-u -- )
      [: { path-addr path-u dir }
        path-addr path-u dir exists? if
          path-addr path-u my-file dir open-file
        else
          path-addr path-u my-file dir create-file
        then
      ;] fat32-tools::current-fs@ with-root-path
    ;

  end-module> import

  \ Receive a file
  : recv-file ( path-addr path-u -- )
    open-or-create-file transfer-data my-file truncate-file
  ;

end-module

File: utils/recv_file.py
#!/usr/bin/env python3

# Copyright (c) 2024 Travis Bemann
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import sys
import serial
import binascii
import base64

PACKET_BODY_SIZE = 512
SEND_PACKET_DONE = (0).to_bytes(4, 'little')
SEND_PACKET_NEXT = (1).to_bytes(4, 'little')
SEND_PACKET_RESEND = (2).to_bytes(4, 'little')
RECV_PACKET_ACK = (3).to_bytes(4, 'little')
RECV_PACKET_NAK = (4).to_bytes(4, 'little')
RECV_PACKET_FAIL = (5).to_bytes(4, 'little')

def make_recv_packet(message):
    len_field = (0).to_bytes(4, 'little')
    packet_minus_crc = message + len_field
    packet = packet_minus_crc + \
             binascii.crc32(packet_minus_crc).to_bytes(4, 'little')
    return base64.b64encode(packet)

RECV_PACKET_ACK_FULL = make_recv_packet(RECV_PACKET_ACK)
RECV_PACKET_NAK_FULL = make_recv_packet(RECV_PACKET_NAK)
RECV_PACKET_FAIL_FULL = make_recv_packet(RECV_PACKET_FAIL)

def generate_send_packet_len():
    body = b'';
    len_field = len(body).to_bytes(4, 'little')
    body_padded = body.ljust(PACKET_BODY_SIZE, b'\x00')
    head = SEND_PACKET_DONE
    packet_minus_crc = head + len_field + body_padded
    packet = packet_minus_crc + \
             binascii.crc32(packet_minus_crc).to_bytes(4, 'little')
    return len(base64.b64encode(packet))

SEND_PACKET_LEN = generate_send_packet_len()

class RecvFile:
    def __init__(self, device, baud):
        self.device = device
        self.baud = baud
        try:
            self.port = serial.Serial(
                port = device,
                baudrate = baud,
                parity = serial.PARITY_NONE,
                stopbits = serial.STOPBITS_ONE,
                bytesize = serial.EIGHTBITS)
            self.port.dtr = True
        except:
            print('Error: TTY device %s invalid' % device)
            sys.exit(1)

    def send_msg(self, data):
        self.port.write(data)
        self.port.flush()

    def recv_data(self, data_file):
        self.send_msg(RECV_PACKET_ACK_FULL)
        done = False
        last_data = None
        while not done:
            packet = base64.b64decode(self.port.read(SEND_PACKET_LEN))
            head = packet[0:4]
            len_field = int.from_bytes(packet[4:8], 'little')
            body = packet[8:8 + min(len_field, PACKET_BODY_SIZE)]
            packet_minus_crc = packet[0:8 + PACKET_BODY_SIZE]
            crc = int.from_bytes(packet[8 + PACKET_BODY_SIZE:
                                        8 + PACKET_BODY_SIZE + 4],
                                 'little')
            if binascii.crc32(packet_minus_crc) == crc:
                if head == SEND_PACKET_DONE:
                    self.send_msg(RECV_PACKET_ACK_FULL)
                    print('Done')
                    done = True
                elif head == SEND_PACKET_NEXT:
                    if last_data != None:
                        data_file.write(last_data)
                    last_data = body
                    self.send_msg(RECV_PACKET_ACK_FULL)
                    print('Received %d bytes' % len_field)
                elif head == SEND_PACKET_RESEND:
                    last_data = body
                    self.send_msg(RECV_PACKET_ACK_FULL)
                    print('Re-received %d bytes' % len_field)
                else:
                    self.send_msg(RECV_PACKET_NAK_FULL)
                    print('Unknown packet')
            else:
                print('CRC doesn\'t match')
                self.send_msg(RECV_PACKET_NAK_FULL)
        if last_data != None:
            data_file.write(last_data)
        self.port.close()

def main():
    if len(sys.argv) != 4:
        print('Usage: %s <device> <baud> <file>' % sys.argv[0])
        sys.exit(1)
    data_file = open(sys.argv[3], 'wb')
    RecvFile(sys.argv[1], int(sys.argv[2])).recv_data(data_file)
    data_file.close()

if __name__ == '__main__':
    main()
